README.txt
1. jQuery dependencies
-----------------------

- By default the complete `jQuery UI library' is included. If any single
  component is included, the `jQuery UI library' is discarded and the
  `jQuery UI core' is included instead.
- Any inserted component requires the `jQuery UI core' which is then
  included automatically.
- Any inserted effect requires the `Effects core' which is then included
  automatically.
- Languages are independent components. By default the complete language
  library `i18n' is included. If any single language is included, `i18n'
  is discarded.


2. Directory structure and file names
-------------------------------------

I decided to restructure the default jQuery UI directory tree in
favor of a better consistency and a uniform extension API to copy
with it.
Of course the disadvantage is a slight overhead in integrating new
versions, but it keeps you independent of any directory structure
rearrangements in future jQuery UI releases. 

The structure of the `versions' directory which contains the jQuery
source files differs from the original structure:

versions
  |
  +- 1.3.2
  |    +- languages
  |    +- modules 
  |         |
  |         +- effects
  |         +- widgets
  |         +- interactions
  |         +- core
  |
  +- themes
  
3. File names
-------------

All jQuery UI library names are cropped to their very base name, which
describes their functionality since the respective category is
represented in the directory structure.
 
4. TypoScript
--------------

plugin {
  tx_ptjqueryui {
    # By default the complete jQuery UI library is included.
    # If any single component is included the jQuery UI monolith is
    # discarded and the jQuery core is included automatically.
    version = 1.3.2
    variant = compressed
    position = head
    
    # Any inserted component requires the `jQuery UI core'
    # which is then included automatically.
    components {
      effects {
	# Any inserted effect requires the `Effects core'
	# which is then included automatically.
	1 = blind
	2 = bounce
	3 = clip
	4 = drop
	5 = explode
	6 = fold
	7 = highlight
	8 = pulsate
	9 = scale
	10 = shake
	11 = slide
	12 = transfer
      }
      interactions {
	1 = draggable
	2 = droppable
	3 = resizable
	4 = selectable
	5 = sortable
      }
      widgets {
	1 = accordion
	2 = datepicker
	3 = dialog
	4 = progressbar
	5 = slider
	6 = tabs
      }
    }
    
    # Languages are independent components
    languages {
      1 = ar.js
      2 = bg.js
      3 = ca.js
      4 = cs.js
      5 = da.js
      6 = de.js
      7 = el.js
      8 = eo.js
      9 = es.js
      10 = fa.js
      11 = fi.js
      12 = fr.js
      13 = he.js
      14 = hr.js
      15 = hu.js
      16 = hy.js
      17 = i18n.js
      18 = id.js
      19 = is.js
      20 = it.js
      21 = ja.js
      22 = ko.js
      23 = lt.js
      24 = lv.js
      25 = ms.js
      26 = nl.js
      27 = no.js
      28 = pl.js
      29 = pt-BR.js
      30 = ro.js
      31 = ru.js
      32 = sk.js
      33 = sl.js
      34 = sq.js
      35 = sr.js
      36 = sr-SR.js
      37 = sv.js
      38 = th.js
      39 = tr.js
      40 = uk.js
      41 = zh-CN.js
      42 = zh-TW.js
    }
  }
}


ChangeLog
see pt_jquery/doc/DevDoc.txt!


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "pt_jqueryui"
#
# Auto generated 22-07-2009 18:24
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'jQuery User Interface',
	'description' => 'User Interface Library for jQuery',
	'category' => 'plugin',
	'author' => 'Joachim Mathes',
	'author_email' => 'mathes@punkt.de',
	'shy' => '',
	'dependencies' => 'pt_tools',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
    'author_company' => 'punkt.de GmbH',
	'version' => '0.0.1dev',
	'constraints' => array(
		'depends' => array(
            'pt_tools' => '0.4.2'
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:7:{s:9:"ChangeLog";s:4:"f619";s:10:"README.txt";s:4:"ee2d";s:12:"ext_icon.gif";s:4:"1bdc";s:17:"ext_localconf.php";s:4:"4967";s:19:"doc/wizard_form.dat";s:4:"9c84";s:20:"doc/wizard_form.html";s:4:"d988";s:31:"pi1/class.tx_ptjqueryui_pi1.php";s:4:"d3f6";}',
);

?>

ext_tables.php
<?php
/**
 * Copyright notice
 * 
 * Copyright (c) 2009 Michael Knoll <knoll@punkt.de>
 * All rights reserved
 * 
 * This script is part of the TYPO3 project. The TYPO3 project is 
 * free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * The GNU General Public License can be found at
 * http://www.gnu.org/copyleft/gpl.html.
 * A copy is found in the textfile GPL.txt and important notices to the license 
 * from the author is found in LICENSE.txt distributed with these scripts.
 *
 *
 * This script is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * This copyright notice MUST APPEAR in all copies of the script!
 */



/**
 * Configuration file for tx_ptjqueryui
 *
 * @package		TYPO3
 * @subpackage	pt_jqueryui
 * @author		Michael Knoll <knoll@punkt.de>
 * @version		$Id:$
 */

if (!defined('TYPO3_MODE')) {
	die ('Access denied.');
} // if (!defined('TYPO3_MODE'))


// Add static TS template for JS-Manager plugin
t3lib_extMgm::addStaticFile(
	$_EXTKEY,
	'typoscript/', '[pt_jquery] JS Manager'
);
?>

ext_localconf.php
<?php
/**
 * Copyright notice
 * 
 * Copyright (c) 2009 Michael Knoll <knoll@punkt.de>
 * All rights reserved
 * 
 * This script is part of the TYPO3 project. The TYPO3 project is 
 * free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * The GNU General Public License can be found at
 * http://www.gnu.org/copyleft/gpl.html.
 * A copy is found in the textfile GPL.txt and important notices to the license 
 * from the author is found in LICENSE.txt distributed with these scripts.
 *
 *
 * This script is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * This copyright notice MUST APPEAR in all copies of the script!
 */



/**
 * Configuration file for pt_jqueryui
 * 
 * @author    Michael Knoll <knoll@punkt.de>
 * @since     2009-10-01
 * @version   $Id:$
 */

if (!defined ('TYPO3_MODE')) {
 	die ('Access denied.');
}


// Add JS Manager to static template
t3lib_extMgm::addPItoST43(
    $_EXTKEY,                                   // Extension key 
    'res/class.tx_ptjqueryui_manager.php',      // File to include (relative to extension root)
    '_manager',                                 // Prefix of included file
    'list_type',                                // Type of inclusion to static template (list_type in this case)
    0);                                         // TODO Activate caching, as this changes only, if TS changes --> cache is deleted
    
?>

class.tx_ptjqueryui.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2009 Joachim Mathes <mathes@punkt.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
/**
 * [CLASS/FUNCTION INDEX of SCRIPT]
 *
 * Hint: use extdeveval to insert/update function index above.
 */

require_once(PATH_tslib.'class.tslib_pibase.php');


/**
 * Plugin 'jQuery UI' for the 'pt_jqueryui' extension.
 *
 * @author        Joachim Mathes <mathes@punkt.de>
 * @since         2009-07-23
 * @package       TYPO3
 * @subpackage    tx_ptjqueryui
 */
class tx_ptjqueryui extends tslib_pibase {
    protected $prefixId      = 'tx_ptjqueryui';           // Same as class name
    protected $scriptRelPath = 'class.tx_ptjqueryui.php'; // Path to this script relative to the extension dir.
    protected $extKey        = 'pt_jqueryui';             // The extension key.
    protected $pi_checkCHash = true;

    /**
     * Holds the JavaScript tags to insert into the page header.
     *
     * @var array    $JSTags
     */
    protected $JSTags = array();

    /**
     * Holds the components configuration.
     *
     * The minimized versions of jQuery [UI] are inserted by default.
     *
     * @var array    $JQueryUIconfiguration
     */
    protected $JQueryUIConfiguration = array(
					     'version'    => '',
					     'jq'     => '',
					     'jqui'   => '',
					     'position'   => '',
					     'components' => array(
								   'effects'      => array(),
								   'interactions' => array(),
								   'widgets'      => array()
								   ),
					     'languages'  => array()
					     );



    /**
     * The main method of the PlugIn
     *
     * @param     string    Content passed from TYPO3
     * @param     array     TypoScript configuration passed from TYPO3
     * @return    string    The content that is displayed on the website
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-23
     */
    public function main($content, $conf) {

	$this->conf = $conf;
	
	$this->getConfiguration();
	$this->getData();
	$content .= $this->setData();

	return $content;
    }
   


    /**
     * Processes and saves the TypoScript configuration of jQuery UI.
     *
     * @return    array    The array of file names which shall be included 
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-23
     */
    protected function getConfiguration() {

	// Check where JavaScript tags shall be positioned:
	//  - head (default)
	//  - body
	if (array_key_exists('position', $this->conf) &&
	    strcasecmp($this->conf['position'], 'body') == 0) {

	    $this->JQueryUIConfiguration['position'] = $this->conf['position'];
	} 

        // Check which version shall be included.
        // Default: latest version.
        $availableVersions = $this->getJQueryUIVersions();
	$this->JQueryUIConfiguration['version'] = $availableVersions[count($availableVersions) - 1];

	if (array_key_exists('version', $this->conf)) {

	    if (strcasecmp($this->conf['version'], 'max') == 0) {
		$this->JQueryUIConfiguration['version'] = $availableVersions[count($availableVersions) - 1];
	    } else {
		
		if (in_array($this->conf['version'], $availableVersions)) {
		    $this->JQueryUIConfiguration['version'] = $this->conf['version'];
		} else {
		    $this->JQueryUIConfiguration['version'] = $availableVersions[count($availableVersions) - 1];
		}

	    }
	}

	// Check which JQuery UI variant shall be used.
	// Fallback: 1. normal
	//           2. minimized
	$possibleVariants = array('normal', 'minimized');
	$availableVariants = $this->getJQueryUIVariants($version);
	$variant = '';

	if (array_key_exists('variant', $this->conf)) {

	    if (array_key_exists($this->conf['variant'], $availableVariants)) {
		$variant = $this->conf['variant'];
		$this->JQeryUIConfiguration['jqui'] = $availableVariants[$this->conf['variant']];
	    }

	}

	if ($variant == '') {

	    foreach ($possibleVariants as $variant) {
		if (array_key_exists($variant, $availableVariants)) {
		    $this->JQeryUIConfiguration['jqui'] = $availableVariants[$variant];
		    break;
		} 
	    }

	}


	// Check which JQuery variant shall be used.
	// Fallback: 1. normal
	//           2. minimized
	$possibleVariants = array('normal', 'minimized');
	$availableVariants = $this->getJQueryUIVariants($version);
	$variant = '';

	if (array_key_exists('variant', $this->conf)) {

	    if (array_key_exists($this->conf['variant'], $availableVariants)) {
		$variant = $this->conf['variant'];
		$this->JQeryUIConfiguration['jq'] = $availableVariants[$this->conf['variant']];
	    }

	}

	if ($variant == '') {

	    foreach ($possibleVariants as $variant) {
		if (array_key_exists($variant, $availableVariants)) {
		    $this->JQeryUIConfiguration['jq'] = $availableVariants[$variant];
		    break;
		} 
	    }

	}


	// Check which components shall be included.
	// Check for:
	//  - Effects
	//  - Interactions
	//  - Widgets
	if (array_key_exists('components.', $this->conf)) {
		
	    // Effects
	    if (array_key_exists('effects.', $this->conf['components.']) && 
		is_array($this->conf['components.']['effects.'])) {
            
		foreach ($this->conf['components.']['effects.'] as $number){
		    $this->JQueryUIConfiguration['components']['effects'][] = $this->conf['components.']['effects.'][$number];
		}
	    }

	    // Interactions
	    if (array_key_exists('interactions.', $this->conf['components.']) && 
		is_array($this->conf['components.']['interactions.'])) {
            
		foreach ($this->conf['components.']['interactions.'] as $number){
		    $this->JQueryUIConfiguration['components']['interactions'][] = $this->conf['components.']['interactions.'][$number];
		}
	    }

	    // Widgets
	    if (array_key_exists('widgets.', $this->conf['components.']) && 
		is_array($this->conf['components.']['widgets.'])) {
            
		foreach ($this->conf['components.']['widgets.'] as $number){
		    $this->JQueryUIConfiguration['components']['widgets'][] = $this->conf['components.']['widgets.'][$number];
		}
	    }

        
	    // Check which languages shall be included.
	    if (array_key_exists('languages.', $this->conf) &&
 		is_array($this->conf['languages.'])) {
            
		foreach ($this->conf['languages.'] as $number){
		    $this->JQueryUIConfiguration['components']['languages'][] = $this->conf['languages.'][$number];
		}
	    }
	}


	// Check which languages shall be included.
        $availableLanguages = $this->getJQueryUILanguages();

	if (array_key_exists('languages.', $this->conf)) {

	    foreach ($this->conf['languages.'] as $language){

		if (array_key_exists($language, $availableLanguages)) {
		    $this->JQueryUIConfiguration['languages'][] = $availableLanguages[$language];
		}

	    }
	}
    }


    
    /**
     * Creates JavaScript HTML tags.
     * 
     * Furthermore file dependencies in the jQuery UI configuration are checked.
     *
     * The algorithm of including and excluding components works as follows:
     *  1. The monolithic jQuery UI package is included by default.
     *  2. The monolithic language package, which consists of all languages is
     *     included by default.
     *  3. If any effect, interaction, widget or the core is included explicitly
     *     by a TypoScript configuration, pt_jqueryui discards the monolithic
     *     packages and includes the jQuery UI core instead.
     *  4. If any language is included explicitly by a TypoScript configuration,
     *     the monolithic language package is discarded.
     *
     * @param     void
     * @return    void
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-27
     */
    protected function getData() {
	
	$data = '';

	// Include jQuery library
	$data .= '<script type="text/javascript" src="' . t3lib_extMgm::siteRelPath('pt_jqueryui') . 'versions/' . $this->JQueryUIConfiguration['version'] . '/' . $this->JQueryUIConfiguration['jq'] . '"></script>' . "\n";
	
	// Check components
	if (empty($this->JQueryUIConfiguration['components']['effects']) &&
	    empty($this->JQueryUIConfiguration['components']['interactions']) &&
	    empty($this->JQueryUIConfiguration['components']['widgets'])) {

	    // Include monolithic jQuery and jQuery UI library.
	    $data .= '<script type="text/javascript" src="' . t3lib_extMgm::siteRelPath('pt_jqueryui') . 'versions/' . $this->JQueryUIConfiguration['version'] . '/' . $this->JQueryUIConfiguration['jqui'] . '"></script>' . "\n";
	}
	else {

	    // Include jQuery components
	    foreach ($this->JQueryUIConfiguration['components'] as $key => $value) {

		if (!empty($value)) {

		    foreach ($value as $component) {
			$data .= '<script type="text/javascript" src="' . t3lib_extMgm::siteRelPath('pt_jqueryui') . 'versions/' . $this->JQueryUIConfiguration['version'] . $this->JQueryUIConfiguration['components'][$key][$component] . '"></script>' . "\n";
		    } 
		}
	    }
	}

	// Check languages
	if (empty($this->JQueryUIConfiguration['languages'])) {
	    
	    // Include monolithic jQuery language library.
	    $data .= '<script type="text/javascript" src="' . t3lib_extMgm::siteRelPath('pt_jqueryui') . 'versions/' . $this->JQueryUIConfiguration['version'] . '/languages/i18n.js"></script>' . "\n";
	}
	else {

	    // Include individual jQuery files
	    foreach ($this->JQueryUIConfiguration['languages'] as $language) {
		$data .= '<script type="text/javascript" src="' . t3lib_extMgm::siteRelPath('pt_jqueryui') . 'versions/' . $this->JQueryUIConfiguration['version'] . '/languages/' . $this->JQueryUIConfiguration['languages'][$language] . '"></script>' . "\n";
	    } 

	}

	$this->data = $data;
    }


    /**
     * Includes JavaScript tags into TYPO3 page.
     *
     * TODO: This method should register all included JavaScript files in a
     *       framework, which takes care of not including conflicting
     *       JavaScripts. The framework could be based on the TYPO3 extension
     *       `jsmanager' by Joerg Schoppet.
     *
     * @param     void
     * @return    string    content
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-24
     */
    protected function setData() {

	$content = '';

        switch($this->position) {
        case 'head':
            $GLOBALS['TSFE']->additionalJavaScript[$this->extKey] = implode("\n", $this->data);
            break;
	case 'body':
	    $content = implode("\n", $this->data);
	    break;
	default:
	    $GLOBALS['TSFE']->additionalJavaScript[$this->extKey] = implode("\n", $this->data);
        }
	
	return $content;
    }


    
    /**
     * Returns available versions of jQuery UI.
     * 
     * To achieve this, it iterates over the versions-directory. It expects the directories
     * to be named like
     * <directory> ::= <number>(.<number>)*
     *
     * return     array    versions
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-23
     */
    protected function getJQueryUIVersions() {

	$versions = array();
	$files = new DirectoryIterator(t3lib_extMgm::extPath('pt_jqueryui', 'versions'));

	foreach ($files as $file) {
	    if ($file->isDir() && !$file->isDot()){
		$versions[] = $file->getFilename();
	    }
	}

	sort($versions);

	return $versions;
    }




    /**
     * Returns available variants of jQuery UI, i.e. normal or minimized.
     *
     * @param     string    Defines the version directory in which to search.
     * @return    array     variants
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-23
     */
    protected function getJQueryUIVariants($version) {

	$possibleVariants = array(
				  'normal' => '.',
				  'minimized' => '.min.',
				  );
	$variants = array();
	$files = new DirectoryIterator(t3lib_extMgm::extPath('pt_jqueryui', 'versions/' . $version));
	
        foreach ($files as $file) {
            if ($file->isFile()) {
                $fileName = $file->getFilename();
		$fileName = str_replace('jquery-ui', '', $fileName);
		$fileName = str_replace('js', '', $fileName);
		$variant = array_search($fileName, $possibleVariants);
		
		if ($variant) {
		    $variants[$variant] = $entry->getFilename();
		}
            }
        }

	return $variants;
    }



    /**
     * Returns available variants of jQuery, i.e. normal or minimized.
     *
     * @param     string    Defines the version directory in which to search.
     * @return    array     variants
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-27
     */
    protected function getJQueryVariants($version) {

	$possibleVariants = array(
				  'normal' => '.',
				  'minimized' => '.min.',
				  );
	$variants = array();
	$files = new DirectoryIterator(t3lib_extMgm::extPath('pt_jqueryui', 'versions/' . $version));
	
        foreach ($files as $file) {
            if ($file->isFile()) {
                $fileName = $file->getFilename();
		$fileName = str_replace('jquery', '', $fileName);
		$fileName = str_replace('js', '', $fileName);
		$variant = array_search($fileName, $possibleVariants);
		
		if ($variant) {
		    $variants[$variant] = $entry->getFilename();
		}
            }
        }

	return $variants;
    }



    /**
     * Returns available languages of jQuery UI.
     *
     * To achieve this, it iterates over the respective version's language
     * directory.
     *
     * @param     string    Defines the version directory in which to search.
     * @return    array     languages
     * @author    Joachim Mathes <mathes@punkt.de>
     * @since     2009-07-24
     */
    protected function getJQueryUILanguages($version) {

	$languages = array();
	$files = new DirectoryIterator(t3lib_extMgm::extPath('pt_jqueryui', 'versions/' . $version . '/languages'));
	
        foreach ($files as $file) {
            if ($file->isFile()) {
		$fileName = $file->getFilename();
		$fileName = str_replace('.js', '', $fileName);
		$languages[$filename] = $file->getFilename();
            }
        }

	return $languages;
    }


}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pt_jqueryui/class.tx_ptjqueryui.php']) {
    include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pt_jqueryui/class.tx_ptjqueryui.php']);
}

?>

typoscript/config.ts
plugin {

  tx_ptjqueryui {
  
    # By default the complete jQuery UI library is included.
    # If any single component is included the jQuery UI monolith is
    # discarded and the jQuery core is included automatically.
    version = 1.3.2
    variant = compressed
    position = head
    
    # Any inserted component requires the `jQuery UI core'
    # which is then included automatically.
    components {
    
      effects {
      
    	# Any inserted effect requires the `Effects core'
    	# which is then included automatically.
    	1 = blind
    	2 = bounce
    	3 = clip
    	4 = drop
    	5 = explode
    	6 = fold
    	7 = highlight
    	8 = pulsate
    	9 = scale
    	10 = shake
    	11 = slide
    	12 = transfer
    	
      }
      
      interactions {
      
    	1 = draggable
    	2 = droppable
    	3 = resizable
    	4 = selectable
    	5 = sortable
    	
      }
      
      widgets {
      
    	1 = accordion
    	2 = datepicker
    	3 = dialog
    	4 = progressbar
    	5 = slider
    	6 = tabs
    	
      }
    }
    
    # Languages are independent components
    languages {
    
      1 = ar
      2 = bg
      3 = ca
      4 = cs
      5 = da
      6 = de
      7 = el
      8 = eo
      9 = es
      10 = fa
      11 = fi
      12 = fr
      13 = he
      14 = hr
      15 = hu
      16 = hy
      18 = id
      19 = is
      20 = it
      21 = ja
      22 = ko
      23 = lt
      24 = lv
      25 = ms
      26 = nl
      27 = no
      28 = pl
      29 = pt-BR
      30 = ro
      31 = ru
      32 = sk
      33 = sl
      34 = sq
      35 = sr
      36 = sr-SR
      37 = sv
      38 = th
      39 = tr
      40 = uk
      41 = zh-CN
      42 = zh-TW
      
    }
    
  }
  
}